pkg/cri/resource-manager/policy/policy.go

```diff
@@ -274,6 +282,89 @@ func (p *policy) ExportResourceData(c cache.Container) {
 	p.cache.WriteFile(c.GetCacheID(), ExportedResources, 0644, buf.Bytes())
 }
 
+// Introspect provides data for external introspection/visualization.
+func (p *policy) Introspect() *introspect.State {
+	pods := p.cache.GetPods()
+	state := &introspect.State{Pods: make(map[string]*introspect.Pod, len(pods))}
+
+	for _, p := range pods {
+		containers := p.GetContainers()
+		if len(containers) == 0 {
+			continue
+		}
+
+		pod := &introspect.Pod{
+			ID:         p.GetID(),
+			UID:        p.GetUID(),
+			Name:       p.GetName(),
+			Containers: make(map[string]*introspect.Container, len(containers)),
+		}
+
+		for _, c := range containers {
+			container := &introspect.Container{
+				ID:      c.GetID(),
+				Name:    c.GetName(),
+				Command: c.GetCommand(),
+				Args:    c.GetArgs(),
+				Hints:   introspect.TopologyHints(c.GetTopologyHints()),
+			}
+			resources := c.GetResourceRequirements()
+			if req, ok := resources.Requests[corev1.ResourceCPU]; ok {
+				if value := req.MilliValue(); value > 0 {
+					container.CPURequest = value
+				}
+			}
+			if lim, ok := resources.Limits[corev1.ResourceCPU]; ok {
+				if value := lim.MilliValue(); value > 0 {
+					container.CPULimit = value
+				}
+			}
+			if req, ok := resources.Requests[corev1.ResourceMemory]; ok {
+				if value := req.Value(); value > 0 {
+					container.MemoryRequest = value
+				}
+			}
+			if lim, ok := resources.Limits[corev1.ResourceMemory]; ok {
+				if value := lim.Value(); value > 0 {
+					container.MemoryLimit = value
+				}
+			}
+			pod.Containers[container.ID] = container
+		}
+		state.Pods[pod.ID] = pod
+	}
+
+	if p.inspsys == nil {
+		sys := &introspect.System{
+			Sockets: make(map[int]*introspect.Socket, p.system.PackageCount()),
+			Nodes:   make(map[int]*introspect.Node, p.system.NUMANodeCount()),
+		}
+		for _, id := range p.system.PackageIDs() {
+			pkg := p.system.Package(id)
+			sys.Sockets[int(id)] = &introspect.Socket{ID: int(id), CPUs: pkg.CPUSet().String()}
+		}
+		for _, id := range p.system.NodeIDs() {
+			node := p.system.Node(id)
+			sys.Nodes[int(id)] = &introspect.Node{ID: int(id), CPUs: node.CPUSet().String()}
+		}
+		sys.Isolated = p.system.Isolated().String()
+		sys.Offlined = p.system.Offlined().String()
+		p.inspsys = sys
+	}
+
+	rdtClassNames := []string{}
+	for _, rdtClass := range rdt.GetClasses() {
+		rdtClassNames = append(rdtClassNames, rdtClass.Name())
+	}
+	p.inspsys.RDTClasses = rdtClassNames
```

Comment:
IMO by including eventually the controller class maps in the introspection data. It will be then up to the visualization UI to decide how to show the case where the set class is not the effective one.

Reply:
> IMO by including eventually the controller class maps in the introspection data. It will be then up to the visualization UI to decide how to show the case where the set class is not the effective one.

This is now tracked separately in #232 .